chia/consensus/blockchain.py

```diff
@@ -1069,61 +1070,62 @@ def seen_compact_proofs(self, vdf_info: VDFInfo, height: uint32) -> bool:
         self._seen_compact_proofs.add(pot_tuple)
         return False
 
-    async def get_block_generator(
-        self, block: BlockInfo, additional_blocks: Optional[Dict[bytes32, FullBlock]] = None
-    ) -> Optional[BlockGenerator]:
-        if additional_blocks is None:
-            additional_blocks = {}
-        ref_list = block.transactions_generator_ref_list
-        if block.transactions_generator is None:
-            assert len(ref_list) == 0
-            return None
-        if len(ref_list) == 0:
-            return BlockGenerator(block.transactions_generator, [])
-
-        result: List[bytes] = []
-        previous_br = await self.get_block_record_from_db(block.prev_header_hash)
-        if previous_br is not None and self.height_to_hash(previous_br.height) == block.prev_header_hash:
-            # We are not in a reorg, no need to look up alternate header hashes
-            # (we can get them from height_to_hash)
-            # in the v2 database, we can look up blocks by height directly
-            # (as long as we're in the main chain)
-            result = await self.block_store.get_generators_at(block.transactions_generator_ref_list)
-        else:
-            # First tries to find the blocks in additional_blocks
-            curr = block
-            additional_height_dict = {}
-            while curr.prev_header_hash in additional_blocks:
-                prev: FullBlock = additional_blocks[curr.prev_header_hash]
-                additional_height_dict[prev.height] = prev
-                curr = prev
-
+    async def lookup_block_generators(self, header_hash: bytes32, generator_refs: Set[uint32]) -> Dict[uint32, bytes]:
+
+        generators: Dict[uint32, bytes] = {}
+
+        # if this is empty, we shouldn't have called this function to begin with
+        assert len(generator_refs)
+
+        # The block heights in the transactions_generator_ref_list don't
+        # necessarily refer to the main chain. The generators may be found in 2
+        # different places. A fork of the chain (but in the database) or in
+        # the main chain.
+
+        #              * <- header_hash
+        #              | :
+        # peak -> *    | : reorg_chain
+        #          \   / :
+        #           \ /  :
+        #            *  <- fork point
+        #         :  |
+        #  main   :  |
+        #  chain  :  |
+        #         :  |
+        #         :  * <- genesis
+
+        # If the block is not part of the main chain, we're on a fork, and we
+        # need to find the fork point
+        peak_block = self.try_block_record(header_hash)
```

Comment:
we can remove the call to try_block_record
as get_block_record_from_db checks the cache before going to db

Reply:
I'm a bit worried about the cost of `await`. Maybe it's an unreasonable concern, but my understanding is that the task is unconditionally suspended when calling `await`, even if the future you're waiting for is ready. i.e. your taks is put at the end of the run queue

Reply:
this suggests that my understanding is wrong though: https://discuss.python.org/t/a-caveat-with-awaiting-futures-that-complete-instantly/37536/5

Reply:
i get the concern but i think it should be ok, we are using async await all along the flow, releasing the event loop, the other tasks that might run during long sync are critical so we dont want long blocks of code without releasing the loop anyway (although from your link it seems that it wont release in the case we have the block cached)

Reply:
I took this suggestion